src/coins.cpp

```diff
@@ -178,80 +189,91 @@ void CCoinsViewCache::SetBestBlock(const uint256 &hashBlockIn) {
     hashBlock = hashBlockIn;
 }
 
-bool CCoinsViewCache::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlockIn, bool erase) {
-    for (CCoinsMap::iterator it = mapCoins.begin();
-            it != mapCoins.end();
-            it = erase ? mapCoins.erase(it) : std::next(it)) {
+bool CCoinsViewCache::BatchWrite(CCoinsMap& mapCoins, CoinsCachePair &sentinel, size_t &childCachedCoinsUsage, const uint256 &hashBlockIn, bool erase) {
+    for (auto it{sentinel.second.Next()}; it != &sentinel; ) {
         // Ignore non-dirty entries (optimization).
-        if (!(it->second.flags & CCoinsCacheEntry::DIRTY)) {
+        if (!(it->second.IsDirty())) {
+            const auto next_entry{it->second.Next(/*clear_flags=*/true)};
+            if (!erase && it->second.coin.IsSpent()) {
+                childCachedCoinsUsage -= it->second.coin.DynamicMemoryUsage();
+                mapCoins.erase(it->first);
```

Comment:
Yes, but if `!erase`. This is what `Sync` was doing before I moved this into `BatchWrite`. Only spent outpoint when we don't want to clear the cache. This is what I already benchmarked to be much faster.

Reply:
This is the reason we need to have the linked list be `CoinsCachePair`s instead of `CCoinsCacheEntry`s, as you mentioned [here](https://github.com/bitcoin/bitcoin/pull/28280#discussion_r1296524385).

Reply:
Err, not `master`, but the previous version of this PR that I force pushed over an hour ago.
In `master` it is looping through the entire `cacheCoins` map, and deleting spent entries. With this approach we do have to do a hash and lookup, but it's constant time instead of linear.

Reply:
2 also required an outpoint based deletion, so this approach just moves that deletion into the same loop instead of clearing unspent flags in the first loop and then looping through spent entries again in a second loop and deleting them via outpoint.

Reply:
If only we could convert a ConsCachePair back into a map iterator...

Reply:
Ok, but is there no C++-fu incantation that can get us from a `std::pair<const COutpoint, CCoinsCacheEntry>*` to a `CCoinsMap::iterator`? I can't figure out a way to get that to compile but I'm sure there's a way using `void*` or something.

Reply:
Can you point me to the way of that ugliness? :pray: